### ISSUE TITLE
Support rangeStrategy=update-lockfile for Bundler when no range in Gemfile

### ISSUE THREAD

**Issue by blischalk** [renovatebot/renovate#7883]
**What Renovate type, platform and version are you using?**

Latest Renovate bot release (23.94.0) in Docker. I have also forked the latest code and replicated this issue with yarn start.

This does not seem to be platform dependent as I have reproduced the issue on GitLab and Github.

**Describe the bug**

Running Renovate against a Ruby code base that does not pin dependencies in its Gemfile and that has upstream updates available in Rubygems does not receive pull requests to update those outdated dependencies.

**To Reproduce**

I have created a simple sample public repo on Github [here](https://github.com/blischalk/renovate-bundler-pinning)

Using a small Gemfile like the following:

```

source 'https://rubygems.org'

group :development do
  gem 'rubocop', '
  gem 'rubocop-performance'
  gem 'rubocop-rspec''
end

```

That has the corresponding Gemfile.lock:

```
GEM
  remote: https://rubygems.org/
  specs:
    ast (2.4.1)
    parallel (1.20.1)
    parser (2.7.2.0)
      ast (~> 2.4.1)
    rainbow (3.0.0)
    regexp_parser (2.0.0)
    rexml (3.2.4)
    rubocop (1.4.2)
      parallel (~> 1.10)
      parser (>= 2.7.1.5)
      rainbow (>= 2.2.2, < 4.0)
      regexp_parser (>= 1.8)
      rexml
      rubocop-ast (>= 1.1.1)
      ruby-progressbar (~> 1.7)
      unicode-display_width (>= 1.4.0, < 2.0)
    rubocop-ast (1.3.0)
      parser (>= 2.7.1.5)
    rubocop-performance (1.9.0)
      rubocop (>= 0.90.0, < 2.0)
      rubocop-ast (>= 0.4.0)
    rubocop-rspec (2.0.0)
      rubocop (~> 1.0)
      rubocop-ast (>= 1.1.0)
    ruby-progressbar (1.10.1)
    unicode-display_width (1.7.0)

PLATFORMS
  ruby

DEPENDENCIES
  rubocop
  rubocop-performance
  rubocop-rspec

BUNDLED WITH
   2.0.1
```

And a config.js like:

```
module.exports = {
  autodiscover: false,
  logLevel: "debug",
  platform: "github",
  gitAuthor: "GitHub Bot <no-replay@foobar.com>",
  branchPrefix: "renovate-",
  onboardingBranch: "renovation-onboarding",
  repositories: [
    {repository: 'blischalk/renovate-bundler-pinning'}
  ]
}
```

Running `yarn start` creates a config PR stating that no updates would be made.

Debug logs also show that there are no available updates for the dependencies that do have upstream updates available:

```
DEBUG: packageFiles with updates (repository=blischalk/renovate-bundler-pinning)
       "config": {
         "bundler": [
           {
             "packageFile": "Gemfile",
             "manager": "bundler",
             "registryUrls": ["https://rubygems.org"],
             "deps": [
               {
                 "depName": "rubocop",
                 "managerData": {"lineNumber": 3},
                 "skipReason": "no-version",
                 "depTypes": ["dvelopment"],
                 "lockedVersion": "1.4.2",
                 "depIndex": 0,
                 "updates": []
               },
               {
                 "depName": "rubocop-performance",
                 "managerData": {"lineNumber": 4},
                 "skipReason": "no-version",
                 "depTypes": ["dvelopment"],
                 "lockedVersion": "1.9.0",
                 "depIndex": 1,
                 "updates": []
               },
               {
                 "depName": "rubocop-rspec",
                 "managerData": {"lineNumber": 5},
                 "skipReason": "no-version",
                 "depTypes": ["dvelopment"],
                 "lockedVersion": "2.0.0",
                 "depIndex": 2,
                 "updates": []
               }
             ],
             "constraints": {"bundler": "2.0.1"}
           }
         ]
       }
```

**What I expect to happen**

Merge requests are opened / would be opened upon config MR merge to upgrade:

Rubocop to ~>1.5.0
Rubocop-performance to ~> 1.9.1
Rubocop-rspec to ~> 2.0.1

**What actually happens**

No merge requests are / would be opened.

**Comment by rarkins** [comment]
I think we should support this if rangeStrategy=update-lockfile ([ref](https://docs.renovatebot.com/configuration-options/#rangestrategy)). The lack of this capability is common across all managers (we tend to ignore updates if there's no range specified, even if there's a lock file).

**Comment by layerssss** [comment]
@rarkins Hi I was gonna ask in a new issue but came across this. Is it true that at the moment I have to specify verison or range in `Gemfile` like this to make it work?

```
gem 'rails', '6.1.1'
```

I've tried it and it does work after adding the version. (I'm using default value of `rangeStrategy` by the way).

I thought if this is required by renovate (pinning version in `Gemfile`) then should there be a "Pin dependencies" PR for `bundler` similarly to `npm`?

**Comment by viceice** [comment]
@layerssss you need lockfile maintenance for lockfile only updates.

Checkout https://docs.renovatebot.com/configuration-options/#lockfilemaintenance

**Comment by layerssss** [comment]
@viceice Thanks for the pointer. It looks like something would work for my current `Gemfile` (with no range for most gems) according to the doc. But I tried it, it didn't create any PR after an immediate renovate run. I thought that's because of what's mentioned in the doc: 

> it defaults its schedule to "before 5am on monday", i.e. to achieve once-per-week semantics.

Anyway I thought I will manually add version range to all gems in `Gemfile` since it would fix it for now. But again thanks for reminding me of this.

**Comment by viceice** [comment]
@layerssss Please open a discussion, so i can help you with your specific config. You shouln'd need to add version ranges.

**Comment by layerssss** [comment]
@viceice I'll be great if that's possible. I've started the discussion here https://github.com/renovatebot/renovate/discussions/8673

**Comment by rarkins** [comment]
To solve this, we need to step through Renovate's logic and "unblock" the concept.

First of all, let's get a reproduction repo with maybe one dependency with a defined Gemfile constraint (needs updating) and another with no constraint but a Gemfile.lock value which needs updating.

From there, we need to make sure that gems without constraints in `Gemfile` are not populated with `skipReason` and *also* that they are populated with `lockedVersion`. `currentValue` can remain `undefined`.

I think that should be enough to get us to https://github.com/renovatebot/renovate/blob/789caadbb49e232ef74f59f5c41c1a6fbffee5c0/lib/workers/repository/process/lookup/index.ts#L29.

I'm not 100% sure what the result will be after that.

**Comment by littldr** [comment]
Sorry I was maybe a bit too eager with starting to implement something.

I created a reproduction repo: https://github.com/littldr/renovate-testing
It contains three gems, one without a version constraint, one with a "normal" version constraint and another one with an open ended constraint (as recommended in the [docs](https://docs.renovatebot.com/ruby/#caveats)). The current versions (in `Gemfile.lock`) are not the latest one.

My idea in #12744 was to mimic Bundlers update behavior. If it finds a gem without a version constraint, it assumes the default range `>= 0` and will update the `Gemfile.lock` to the latest version of this gem with `bundle update`. 

But maybe I'm on the wrong path with my idea? If so, tell me and I'm happy to discuss and contribute a better solution. :)

**Comment by rarkins** [comment]
Everything about the idea is fine, it's more that setting currentValue artificially won't work. What should work is adding a few if/else statements around the place so that we can internally treat an empty/non-existing constraint as if it means "any". First step is to modify bundler/extract to *not* skip non-constrained dependencies plus make sure we've recorded their lockedVersion.

**Comment by littldr** [comment]
ah okay, I think I understand.

Extracting the `lockedVersion` works already for non-constrainted dependencies. `Gemfile.lock` contains the currently used version and the existing code extracts it as `lockedVersion`, so nothing to do here.

I changed #12744 to not skip non-constrained dependencies (and to enforce `rangeStrategy=update-lockfile`, because nothing else makes sense in this case). With this change `currentValue` will be undefined.

The rest of the code skips updating if `currentValue` is undefined and needs some kind of version to compare against. That's why I had to add some "use `lockedVersion` instead of `currentValue`" bits to make the existing structure work.

With this changes I could create a [nice looking PR](https://github.com/littldr/renovate-testing/pull/7) for the `rake` gem in the reproduction repo I posted above. :)

What do you think @rarkins ? Is this the right direction?

**Comment by rarkins** [comment]
We don't want to set `rangeStrategy=update-lockfile` automatically. Doing so would be inconsistent with the way things are in every other package manager. We may one day *default* to `update-lockfile` in general but in the meantime we don't want this.

**Comment by littldr** [comment]
Of course, consistency makes sense. And adjusting the docs that setting `rangeStrategy=update-lockfile` is necessary to get updates for gems without version constraint is good enough.

I now digged a bit deeper into the code and reached a point where I  don't know if proceeding makes sense. To make my example repo work, I replaced at some spots `currentValue` with `currentValue || lockedVersion`. This results in a working update if `rangeStrategy=update-lockfile`, but due to "faking" `currentValue` the rest of the code is not considering this update as a lockfile update anymore (e.g. `isLockfileUpdate` will be `false` in [/lib/workers/repository/process/lookup/generate.ts#L64](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/process/lookup/generate.ts#L64)) and I cannot overlook the impact on the rest of the codebase.

I thought about making `currentValue` optional in all places, but this would involve touching all manager implementations and the change will be huge. Maybe I'm overthinking here? 🤔

**Comment by rarkins** [comment]
I've been able to play with it and got a promising result: https://github.com/renovate-reproductions/3929/pull/2

It's not too many lines of code so will PR it within the next day or two

**Comment by igas** [comment]
I have not tested it yet, but I'm keen to see this use-case covered: https://github.com/littldr/renovate-testing/pull/10/files

**Comment by rarkins** [comment]
@igas please create a new discussion describing your use case.
